### PR TITLE
feat: Checkpoint database when completing an AlephBFT session

### PR DESCRIPTION
**Pull request by m1sterc001guy**
Fixes: https://github.com/fedimint/fedimint/issues/5450

I think this might be useful for federations that potentially corrupt their database. The checkpoint retention  is configurable, so this feature can be turned off if guardian operators don't want the backups.

### CODE REVIEW

File: fedimint-server/src/consensus/engine.rs [hunk 1]
```diff
@@ -472,6 +480,79 @@ impl ConsensusEngine {
...
+        let checkpoint_frequency: u64 = env::var(FM_DB_CHECKPOINT_FREQ_ENV)
+            .unwrap_or(10.to_string())
+            .parse()
+            .expect("FM_DB_CHECKPOINT_FREQ var is invalid");
+
+        if session_index % checkpoint_frequency == 0 {
+            let checkpoint_dir = self.data_dir.join("db_checkpoints");
+
+            if !checkpoint_dir.exists() {
+                if let Err(e) = fs::create_dir_all(&checkpoint_dir) {
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:498]: Would make it once somewhere, unconditionally to avoid having to check over and over.

File: fedimint-server/src/consensus/engine.rs [hunk 2]
```diff
@@ -472,6 +480,79 @@ impl ConsensusEngine {
...
+                    warn!(target: LOG_CONSENSUS, ?checkpoint_dir, ?e, "Could not create checkpoint directory");
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:499]: This should always be possible, we should just fail loudly if we can't do it.

File: fedimint-server/src/consensus/engine.rs [hunk 3]
```diff
@@ -472,6 +480,79 @@ impl ConsensusEngine {
...
+            .expect("FM_DB_CHECKPOINT_SESSION_DIFFERENCE var is invalid");
+        let session_threshold_index = session_index.saturating_sub(session_diff);
+
+        for entry in fs::read_dir(checkpoint_dir)? {
+            let entry = entry?;
+            let session_idx: u64 = entry
+                .file_name()
+                .into_string()
+                .map_err(|_| anyhow!("Could not convert OsString to String"))?
+                .parse()?;
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:547]: Probably better to skip entries that don't parse to a number, just in case some external tool etc. left some trash in there.

File: fedimint-server/src/consensus/engine.rs [hunk 4]
```diff
@@ -472,6 +480,79 @@ impl ConsensusEngine {
...
+            if !checkpoint_dir.exists() {
+                if let Err(e) = fs::create_dir_all(&checkpoint_dir) {
+                    warn!(target: LOG_CONSENSUS, ?checkpoint_dir, ?e, "Could not create checkpoint directory");
+                    return;
+                }
+            }
+
+            let session_checkpoint_dir = checkpoint_dir.join(format!("{session_index}"));
+
+            match self.db.checkpoint(&session_checkpoint_dir) {
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:506]: Wrap in a `TimeReporter` so we can check how long it takes. Based on that we can judge how much a slowdown it is.

File: fedimint-server/src/consensus/engine.rs [hunk 5]
```diff
@@ -472,6 +480,79 @@ impl ConsensusEngine {
...
+        checkpoint_dir: &Path,
+    ) -> anyhow::Result<u64> {
+        // By default, cleanup any checkpoint that is older than 50 sessions
+        let session_diff: u64 = env::var(FM_DB_CHECKPOINT_SESSION_DIFFERENCE_ENV)
+            .unwrap_or(50.to_string())
+            .parse()
+            .expect("FM_DB_CHECKPOINT_SESSION_DIFFERENCE var is invalid");
+        let session_threshold_index = session_index.saturating_sub(session_diff);
+
+        for entry in fs::read_dir(checkpoint_dir)? {
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:541]: Instead of iterating every time, we could use some more adaptive strategy.

E.g. we could generate a random number between 0 and session different, and if it exist, delete it. This would prune most checkpoints, leaving out some snapshot that could be used to faster sync. Wrapping the whole in a TimeReporter probably a good idea too.

Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs:541]: Just realized that we don't need to iterate at all. Since we know the frequency and the current session, we should know the index of the session that needs to be deleted. Can just check if that exists, delete it, and be done.

**Comment by m1sterc001guy** [comment]
Confirmed that checkpoints are incremental, since they take advantage of Rocksdb's internal structure to avoid doing a heavy copy each checkpoint: https://rocksdb.org/blog/2015/11/10/use-checkpoints-for-efficient-snapshots.html

TLDR: A hard link will be made for SSTables that are the same as the current database and a full copy is made for differing files. Manifest and log are copied each checkpoint.

File: fedimint-server/src/consensus/engine.rs [hunk 6]
```diff
@@ -472,6 +484,81 @@ impl ConsensusEngine {
...
+    }
+
+    /// Creates a backup of the database in the checkpoint directory. The
+    /// frequency of the checkpoints is controlled by
+    /// `FM_DB_CHECKPOINT_FREQ_ENV`. These checkpoints can be used to
+    /// restore the database in case the federation falls out of consensus.
+    fn checkpoint_database(&self, session_index: u64) {
+        // By default, checkpoint the database every 10 sessions
+        let checkpoint_frequency: u64 = env::var(FM_DB_CHECKPOINT_FREQ_ENV)
+            .unwrap_or(10.to_string())
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:506]: I would default to `1`. Seems like a safest option.

`0` or `off` should turn it off completely, in case it turns out too slow.

File: fedimint-server/src/consensus/engine.rs [hunk 7]
```diff
@@ -472,6 +484,81 @@ impl ConsensusEngine {
...
+        Ok(())
+    }
+
+    /// Creates a backup of the database in the checkpoint directory. The
+    /// frequency of the checkpoints is controlled by
+    /// `FM_DB_CHECKPOINT_FREQ_ENV`. These checkpoints can be used to
+    /// restore the database in case the federation falls out of consensus.
+    fn checkpoint_database(&self, session_index: u64) {
+        // By default, checkpoint the database every 10 sessions
+        let checkpoint_frequency: u64 = env::var(FM_DB_CHECKPOINT_FREQ_ENV)
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:505]: Moving reading this to a new function, and in some previous performance-sensitive code, I used `Once` to make sure it's as fast as possible.

File: fedimint-server/src/consensus/engine.rs [hunk 8]
```diff
@@ -472,6 +484,81 @@ impl ConsensusEngine {
...
+            .unwrap_or(10.to_string())
+            .parse()
+            .expect("FM_DB_CHECKPOINT_FREQ var is invalid");
+
+        if session_index % checkpoint_frequency == 0 {
+            let checkpoint_dir = self.data_dir.join("db_checkpoints");
+            let session_checkpoint_dir = checkpoint_dir.join(format!("{session_index}"));
+
+            {
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:514]: Thought: it might be asking for trouble, but these two should be able to run in parallel.

File: fedimint-server/src/consensus/engine.rs [hunk 9]
```diff
@@ -472,6 +484,81 @@ impl ConsensusEngine {
...
+                if let Err(e) =
+                    Self::delete_old_database_checkpoints(session_index, &checkpoint_dir)
+                {
+                    warn!(target: LOG_CONSENSUS, ?e, "Could not delete old checkpoints");
+                }
+            }
+        }
+    }
+
+    /// Iterates through the db checkpoint directory and deletes the checkpoint
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:538]: No longer iterates.

File: fedimint-server/src/consensus/engine.rs [hunk 10]
```diff
@@ -472,6 +484,81 @@ impl ConsensusEngine {
...
+    /// equal to `session_index` - `session_diff`.
+    fn delete_old_database_checkpoints(
+        session_index: u64,
+        checkpoint_dir: &Path,
+    ) -> anyhow::Result<()> {
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:543]: Thought: should we do the iteration one on start, just to avoid leftovers when changing retention settings?

File: fedimint-server/src/envs.rs [hunk 11]
```diff
@@ -1,3 +1,10 @@
 /// The env var for maximum open connections the API can handle
 pub const FM_MAX_CLIENT_CONNECTIONS_ENV: &str = "FM_MAX_CLIENT_CONNECTIONS";
 pub const FM_PEER_ID_SORT_BY_URL_ENV: &str = "FM_PEER_ID_SORT_BY_URL";
+
+/// Environment variable for the session count determining when to cleanup old
+/// checkpoints.
+pub const FM_DB_CHECKPOINT_SESSION_DIFFERENCE_ENV: &str = "FM_DB_CHECKPOINT_SESSION_DIFFERENCE";
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply fedimint-server/src/envs.rs]: Removed the frequency one and renamed the other to `FM_DB_CHECKPOINT_RETENTION`

File: fedimint-server/src/consensus/engine.rs (hunk 10 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs:558]: Yes, good idea. Added to `initialize_checkpoint_directory`

File: fedimint-server/src/consensus/engine.rs (hunk 9 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs]: Removed

File: fedimint-server/src/consensus/engine.rs (hunk 6 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs]: Removed entirely, always checkpoint ever session now.

File: fedimint-server/src/consensus/engine.rs (hunk 7 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs]: Moved to `run`

File: fedimint-server/src/consensus/engine.rs (hunk 8 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs]: I'm leaving this for now, I think it will be fast enough without it.

File: fedimint-rocksdb/src/lib.rs [hunk 12]
```diff
@@ -165,6 +171,10 @@ impl IRawDatabase for RocksDbReadOnly {
     async fn begin_transaction<'a>(&'a self) -> RocksDbReadOnlyTransaction<'a> {
         RocksDbReadOnlyTransaction(&self.0)
     }
+
+    fn checkpoint(&self, _backup_path: &Path) -> Result<()> {
+        Err(anyhow!("Cannot checkpoint a read only database"))
```
Comment (dpc) [inline fedimint-rocksdb/src/lib.rs:176]: Thought: why not?

File: fedimint-server/src/consensus/engine.rs [hunk 13]
```diff
@@ -472,6 +484,89 @@ impl ConsensusEngine {
...
+            debug!(
+                ?current_session,
+                "Removing database checkpoints up to `current_session`"
+            );
+
+            for checkpoint in fs::read_dir(checkpoint_dir)?.flatten() {
+                // Validate that the directory is a session index
+                if let Ok(file_name) = checkpoint.file_name().into_string() {
+                    if let Ok(session) = file_name.parse::<u64>() {
+                        if session < current_session - self.checkpoint_retention {
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:503]: Possibility of  underflow in debug node?

File: fedimint-server/src/consensus/engine.rs [hunk 14]
```diff
@@ -472,6 +484,89 @@ impl ConsensusEngine {
             .expect("This is the only place where we write to this key");
     }
 
+    /// Creates the directory within the data directory for storing the database
+    /// checkpoints or deletes checkpoints before `current_session` -
+    /// `checkpoint_retention`.
+    fn initialize_checkpoint_directory(&self, current_session: u64) -> anyhow::Result<()> {
+        let checkpoint_dir = self.data_dir.join("db_checkpoints");
```
Comment (dpc) [inline fedimint-server/src/consensus/engine.rs:491]: Nit: would make a `fn db_checkpoints_dir() -> PathBuf` to avoid possibility of typos

Comment (dpc) [reply fedimint-server/src/consensus/engine.rs:491]: And maybe `fn db_checkpoint_dir(session_num: u64) -> PathBuf` maybe. Seems like it repeats couple of times.

File: fedimint-server/src/consensus/mod.rs [hunk 15]
```diff
@@ -131,6 +138,11 @@ pub async fn run(
         );
     }
 
+    let checkpoint_retention: u64 = env::var(FM_DB_CHECKPOINT_RETENTION_ENV)
+        .unwrap_or(FM_DB_CHECKPOINT_RETENTION_DEFAULT.to_string())
+        .parse()
+        .expect("FM_DB_CHECKPOINT_SESSION_DIFFERENCE var is invalid");
```
Comment (dpc) [inline fedimint-server/src/consensus/mod.rs:144]: Previous name still here? Possibly a sign to use the const here too.

**Review by dpc — COMMENTED** [review]
Landable, but enough nits and suggestions so not landing just yet.

File: fedimint-server/src/consensus/engine.rs (hunk 13 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs]: Good catch, added check

File: fedimint-rocksdb/src/lib.rs (hunk 12 above)
Comment (m1sterc001guy) [reply fedimint-rocksdb/src/lib.rs]: Added :)

File: fedimint-server/src/consensus/engine.rs (hunk 14 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/engine.rs]: Done

File: fedimint-server/src/consensus/mod.rs [hunk 16]
```diff
@@ -33,18 +35,23 @@ use tracing::log::warn;
...
 use crate::consensus::engine::ConsensusEngine;
+use crate::envs::FM_DB_CHECKPOINT_RETENTION_ENV;
 use crate::net;
 use crate::net::api::{ApiSecrets, RpcHandlerCtx};
 
 /// How many txs can be stored in memory before blocking the API
 const TRANSACTION_BUFFER: usize = 1000;
 
+// How many checkpoints from the current session should be retained on disk.
+const FM_DB_CHECKPOINT_RETENTION_DEFAULT: u64 = 1;
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/mod.rs]: Done

File: fedimint-server/src/consensus/mod.rs (hunk 15 above)
Comment (m1sterc001guy) [reply fedimint-server/src/consensus/mod.rs]: Fixed, added constant as well

**Review by dpc — APPROVED** [review]

**Review by dpc — APPROVED** [review]
